### ISSUE TITLE
Shuffle=True

### ISSUE THREAD

**Issue by summa-code** [mlech26l/learning-long-term-irregular-ts#4]
I saw this line,

```trainloader = data.DataLoader(train, batch_size=64, shuffle=True, num_workers=4)```

for timeseries, should shuffle be True ? Just curious, i have not shuffled the data in sliding window models,
And also these lines are interesting, you really do shuffle the data.

```
permutation = np.random.RandomState(98841).permutation(total_seqs)
        test_size = int(0.2 * total_seqs)
        self.test_x = all_x[permutation[:test_size]]
        self.test_y = all_y[permutation[:test_size]]
        self.test_t = all_t[permutation[:test_size]]
        self.train_x = all_x[permutation[test_size:]]
        self.train_t = all_t[permutation[test_size:]]
        self.train_y = all_y[permutation[test_size:]]
```
I am student still learning. :-)

**Comment by mlech26l** [comment]
Hi, 
the training data array has 3 axes: [samples, timesteps, features] and the DataLoader shuffles along the first axis, i.e., shuffles the ordering of the sequences. The shuffling is required to make sure we are performing **stochastic** gradient descent, i.e., during each epoch iterating over the training data by a different ordering than the ordering of the previous epochs.

The code block you are quoting is not about the training process but the test-training split. When we collect the data we might collect the samples of class A first, then the samples of class B and so on. Later when we split the data into training-validation-test set we have to shuffle the data first to ensure the samples of class A,B,... between the splits are uniformly distributed. If we wouldn't shuffle here, then our test set would consist only of samples of class A, whereas no class A samples would be in the training set.

**Comment by summa-code** [comment]
Oh ok. I understood about shuffling of the ordering. I have another question, I am asking these questions because i am dealing with Time series data, and the time is uneven, not equidistant. For example, a data feed about stock symbol and the volume and price. I see that the time component is removed and only the time difference between two data points are taken.

```             
                if last_millis is None:
                    elasped_sec = 0.05
                else:
                    elasped_sec = float(millis - last_millis) / 1000.0
                elasped = elasped_sec * 1000 / millis_mapped_to_1
```
I have seen in places where they convert time into a frequency component, such as time of day, day of year/month (cos/sin). I am just curious to learn more about ODE and how that can be applied to uneven timeseries data involving timestamp. Another example is Urban activity and measurements of air quality samples, predicting few time steps into the future.

And most of the work is done by shifting the label by timesteps into the future in the training data. I am not sure if the same can be applied here.

**Comment by mlech26l** [comment]
There are several approaches on how to deal with irregularly sampled time-series. For instance, the simplest solution is to just treat the elapsed-time input as an additional input feature and then use a standard sequence processing model such as an LSTM. This works actually quite well, but other solutions that add an [inductive bias](https://en.wikipedia.org/wiki/Inductive_bias) for the time input to the model perform even better.  

One of such options is to transform the time into some spectral domain as you suggested. The [Phased-LSTM](https://papers.nips.cc/paper/2016/file/5bce843dd76db8c939d5323dd3e54ec9-Paper.pdf) is doing exactly that.
Another method is to use ODEs: Parametrized the ODE by a neural network and use the time input as the solution time of the ODE-solver. 

The elapsed-time vs total-passed-time is just a difference in the implementation if you have to solve the ODE from [0,elapsed_time] for [previous-total-time, total-time].

**Comment by summa-code** [comment]
Oh i tried LSTM taking time as frequency component (spectral domain). It was terrible. The problem could be because of the shifting of label values (pre-trained train data). They call it windowing and label shift ( t + n ). I followed what google's team suggested. The performance was really poor/bad. May be i should not shift the label, but predict the values based on the time as input for varies time steps. I will look into other models as well. You can see here as to what i mean by label shift. This is very basic implementation from google.
https://www.tensorflow.org/tutorials/structured_data/time_series#data_windowing
The performance is really poor. I am still not sure of the phase shift of label data. I understand the sequencing into the future. But i am little bit lost in that implementation. 

With respect to ODE, do you think that i should consider total time instead of elapsed time ? I can try spectral domain, not sure how would work. I see one feature component here

```series_t.append(elasped)```

I am trying to understand your statement there for this one ```[previous-total-time, total-time]``` And also time is linearly increasing value like an entropy. This would present problems when we normalize time component. I have gone through Phased LSTM once before, and there are couple of implementations out there. My bad you have implemented PhasedLSTM. I did not notice. I edited my comment and removed my ignorance. I will try your code. I see that the same elapsed time is considered here.
```        if (isinstance(inputs, tuple) or isinstance(inputs, list)) and len(inputs) > 1:
            elapsed = inputs[1]  
```
I thought ODE is little different and could be better compared to Phased LSTM. But it is interesting challenge to figure out best algorithm for Time related problems and to predict something into the future that it had not encountered before.

**Comment by mlech26l** [comment]
If you get really bad results with a simple LSTM + time as an additional input baseline, I expect that there is something wrong with the data. For instance, not enough data, bad preprocessing (not normalized features), or maybe you are facing a problem that is difficult to solve in general (think of predicting random noise, which is impossible).

All the fancy time-continuous RNNs (e.g., ODE, Phased-LSTM, CT-GRU) give you a bit better accuracy than the baseline but I wouldn't expect magic improvement. If your baseline doesn't work the problem probably lies somewhere else.

**Comment by summa-code** [comment]
And yes, you are right. This problem is little complex. The data is a regular feed, and it is pretty good from what i have seen. Your code is concise. And Thanks for publishing your implementation. There is so much that i could do in preprocessing and trying different models. Most of all, i appreciate your inputs. Thank you very much. :pray:
